### PR TITLE
Change `TrainingDataImporter` to be sync

### LINKED ISSUE #9131: Change `TrainingDataImporter` to be sync
**Description of Problem**:
Currently our `TrainingDataImporter` interface (and implementations) has a lot of `async` methods.
This will make integrating it into the new graph architecture difficult + we don't actually make sure of any async IO calls (as far as I'm aware).

**Overview of the Solution**:
We need to make all the methods in the importer classes normal, not `async` methods.

**Definition of Done**:
- [ ] Change all the interface and all importers to have only `sync` methods.
- [ ] Update the related tests to also not be `async`.
- [ ] Docs/migration guide for custom importers.

### PR DESCRIPTION
**Pull request by alwx**
**Proposed changes**:
- fix #9131 

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: changelog/9131.misc.md [hunk 1]
```diff
@@ -0,0 +1 @@
+Change `TrainingDataImporter` to be sync.
```
Comment (wochinge) [inline changelog/9131.misc.md:1]: Can we please also update the docs [here](https://rasa.com/docs/rasa/training-data-importers) and add a migration guide entry? 🙌🏻

File: rasa/core/test.py [hunk 2]
```diff
@@ -428,9 +428,9 @@ async def _create_data_generator(
         training_data_paths=[resource_name], domain_path=domain_path
     )
     if use_conversation_test_files:
-        story_graph = await test_data_importer.get_conversation_tests()
+        story_graph = test_data_importer.get_conversation_tests()
```
Comment (wochinge) [inline rasa/core/test.py:431]: `_create_data_generator` doesn't need to. be `async` anymore, right?

File: rasa/model.py [hunk 3]
```diff
@@ -329,6 +329,7 @@ def project_fingerprint() -> Optional[Text]:
         return None
 
 
+# TODO(alwx): doesn't need to be async
```
Comment (wochinge) [inline rasa/model.py:332]: open todo

File: rasa/model_training.py [hunk 4]
```diff
@@ -516,9 +512,8 @@ async def _train_core_with_validated_data(
         rasa.shared.utils.cli.print_color(
             "Training Core model...", color=rasa.shared.utils.io.bcolors.OKBLUE
         )
-        domain, config = await asyncio.gather(
-            file_importer.get_domain(), file_importer.get_config()
-        )
+        domain = file_importer.get_domain()
```
Comment (wochinge) [inline rasa/model_training.py:515]: `_core_model_for_finetuning` also doesn't need to be `async` once the fingerprinting method isn't async any longer

File: rasa/nlu/test.py [hunk 5]
```diff
@@ -1409,6 +1409,7 @@ def remove_pretrained_extractors(pipeline: List[Component]) -> List[Component]:
     return pipeline
 
 
+# TODO(alwx): no async?
```
Comment (wochinge) [inline rasa/nlu/test.py:1412]: TODO

File: rasa/shared/core/training_data/loading.py [hunk 6]
```diff
@@ -75,7 +75,7 @@ async def load_data_from_resource(
     if not os.path.exists(resource):
         raise ValueError(f"Resource '{resource}' does not exist.")
 
-    return await load_data_from_files(
+    return load_data_from_files(
```
Comment (wochinge) [inline rasa/shared/core/training_data/loading.py:78]: can be `sync` then, right?

File: rasa/shared/importers/utils.py [hunk 7]
```diff
@@ -12,17 +12,17 @@ def training_data_from_paths(paths: Iterable[Text], language: Text) -> TrainingD
...
-async def story_graph_from_paths(
+def story_graph_from_paths(
     files: List[Text],
     domain: Domain,
     template_variables: Optional[Dict] = None,
     use_e2e: bool = False,
     exclusion_percentage: Optional[int] = None,
 ) -> StoryGraph:
-
+    """Returns the StoryGraph from paths."""
```
Comment (wochinge) [inline rasa/shared/importers/utils.py:22]:
```suggestion
    """Returns the `StoryGraph` from paths."""
```

File: tests/shared/importers/test_importer.py [hunk 8]
```diff
@@ -50,7 +50,7 @@ async def test_use_of_interface():
     ]
     for f in functions_to_test:
         with pytest.raises(NotImplementedError):
-            await f()
+            f()
 
 
 async def test_combined_file_importer_with_single_importer(project: Text):
```
Comment (wochinge) [inline tests/shared/importers/test_importer.py:56]:
```suggestion
def test_combined_file_importer_with_single_importer(project: Text):
```

Comment (wochinge) [reply tests/shared/importers/test_importer.py:56]: there are a couple of more `async`s in the module btw

**Review by wochinge — CHANGES_REQUESTED** [review]
Wow! Great work tracing all the no longer required `async`s 💯 

Does this have implications onto Rasa X?

File: docs/docs/training-data-importers.mdx [hunk 9]
```diff
@@ -212,3 +212,9 @@ class MyImporter(TrainingDataImporter):
     def _custom_get_nlu_file(self) -> Text:
         pass
 ```
+
+## Migrating your Custom Importer to Rasa 3.0
+
+`TrainingDataImporter` and its subclasses contain no async methods since Rasa 3.0.
+In order to migrate your custom importers and make them work with Rasa 3.0, you also need to
+replace your async methods with synchronised ones.
```
Comment (alwx) [inline docs/docs/training-data-importers.mdx:220]: @wochinge I honestly don't have any ideas of what else to write here. Should there be an example or is it straightforward enough already?

File: tests/core/test_policies.py [hunk 10]
```diff
@@ -261,7 +257,7 @@ async def test_finetune_after_load(
                 ActionExecuted(ACTION_LISTEN_NAME),
             ],
         )
-        original_train_data = await train_trackers(
+        original_train_data = train_trackers(
```
Comment (wochinge) [inline tests/core/test_policies.py:260]: the test fn can be sync then, right?

File: tests/nlu/test_evaluation.py [hunk 11]
```diff
@@ -386,7 +384,7 @@ async def test_eval_data(
     )
     interpreter = Interpreter.load(nlu_model_directory, component_builder)
 
-    data = await data_importer.get_nlu_data()
+    data = data_importer.get_nlu_data()
```
Comment (wochinge) [inline tests/nlu/test_evaluation.py:387]: the test fn doesn't need `async` anymore

File: tests/shared/importers/test_importer.py [hunk 12]
```diff
@@ -311,21 +309,21 @@ async def test_adding_e2e_actions_to_domain(default_importer: E2EImporter):
...
 
-    async def mocked_stories(*_: Any, **__: Any) -> StoryGraph:
+    def mocked_stories(*_: Any, **__: Any) -> StoryGraph:
         return stories
 
     # Patch to return our test stories
     default_importer.importer.get_stories = mocked_stories
 
-    domain = await default_importer.get_domain()
+    domain = default_importer.get_domain()
```
Comment (wochinge) [inline tests/shared/importers/test_importer.py:318]: the test fn doesn't need `async` anymore

File: tests/shared/importers/test_rasa.py [hunk 13]
```diff
@@ -22,20 +22,20 @@ async def test_rasa_file_importer(project: Text):
 
     importer = RasaFileImporter(config_path, domain_path, [default_data_path])
 
-    domain = await importer.get_domain()
+    domain = importer.get_domain()
```
Comment (wochinge) [inline tests/shared/importers/test_rasa.py:25]: the test fns don't need `async` anymore

File: tests/shared/nlu/training_data/test_training_data.py [hunk 14]
```diff
@@ -744,11 +743,11 @@ async def test_without_additional_e2e_examples(tmp_path: Path):
     )
 
     # Patch to return our test stories
-    existing.get_stories = asyncio.coroutine(lambda *args: stories)
+    existing.get_stories = lambda *args: stories
 
     importer = E2EImporter(existing)
 
-    training_data = await importer.get_nlu_data()
+    training_data = importer.get_nlu_data()
```
Comment (wochinge) [inline tests/shared/nlu/training_data/test_training_data.py:750]: the test fn doesn't need `async` anymore

File: tests/test_model.py [hunk 15]
```diff
@@ -582,12 +581,12 @@ async def test_update_with_new_domain(trained_rasa_model: Text, tmpdir: Path):
...
     mocked_importer = Mock()
 
-    async def get_domain() -> Domain:
+    def get_domain() -> Domain:
         return new_domain
 
     mocked_importer.get_domain = get_domain
 
-    await model.update_model_with_new_domain(mocked_importer, tmpdir)
+    model.update_model_with_new_domain(mocked_importer, tmpdir)
```
Comment (wochinge) [inline tests/test_model.py:589]: the test fn doesn't need `async` anymore

File: tests/test_model.py [hunk 16]
```diff
@@ -607,12 +606,12 @@ async def test_update_with_new_domain_preserves_domain(
...
     mocked_importer = Mock()
 
-    async def get_domain() -> Domain:
+    def get_domain() -> Domain:
         return Domain.load(domain_with_categorical_slot_path)
 
     mocked_importer.get_domain = get_domain
 
-    await model.update_model_with_new_domain(mocked_importer, tmpdir)
+    model.update_model_with_new_domain(mocked_importer, tmpdir)
```
Comment (wochinge) [inline tests/test_model.py:614]: the test fn doesn't need `async` anymore

**Comment by alwx** [comment]
@wochinge updated all the docs and made it all look much nicer than before. Please, check again!

File: docs/docs/migration-guide.mdx [hunk 17]
```diff
@@ -10,6 +10,63 @@ description: |
 This page contains information about changes between major versions and
 how you can migrate from one version to another.
 
+## Rasa 2.x to 3.0
```
Comment (wochinge) [inline docs/docs/migration-guide.mdx:13]: @melindaloubser1 Could you please give this a short review?

File: docs/docs/migration-guide.mdx [hunk 18]
```diff
@@ -10,6 +10,63 @@ description: |
...
+    def _custom_get_domain_file(self) -> Text:
+        pass
+
+    # this function was also async before
+    def get_config(self) -> Dict:
+        path_to_config = self._custom_get_config_file()
+        return rasa.utils.io.read_config_file(path_to_config)
+
+    def _custom_get_config_file(self) -> Text:
+        pass
```
Comment (wochinge) [inline docs/docs/migration-guide.mdx:65]: same as above
```suggestion
        pass
```

File: docs/docs/migration-guide.mdx [hunk 19]
```diff
@@ -10,6 +10,63 @@ description: |
...
+class MyImporter(TrainingDataImporter):
+    """Example partial implementation of a custom importer component."""
+
+    # this function was async before
+    def get_domain(self) -> Domain:
+        path_to_domain_file = self._custom_get_domain_file()
+        return Domain.load(path_to_domain_file)
+
+    def _custom_get_domain_file(self) -> Text:
+        pass
```
Comment (wochinge) [inline docs/docs/migration-guide.mdx:57]: optional: how about shortening that part as it's not of interest for this example?

```suggestion
        pass

    def _custom_get_domain_file(self) -> Text:
        pass
```

File: docs/docs/training-data-importers.mdx [hunk 20]
```diff
@@ -33,6 +33,13 @@ importers:
 The `name` key is used to determine which importer should be loaded. Any extra
 parameters are passed as constructor arguments to the loaded importer.
 
+:::note
+`TrainingDataImporter` and its subclasses contain no async methods since Rasa 3.0.
+In order to migrate your custom importers and make them work with Rasa 3.0, you also need to
+replace your async methods with synchronised ones.
```
Comment (wochinge) [inline docs/docs/training-data-importers.mdx:39]:
```suggestion
replace your async methods with synchronized ones.
```

File: rasa/core/train.py [hunk 21]
```diff
@@ -108,17 +108,15 @@ async def train_comparison_models(
...
-                        train(
-                            domain,
-                            file_importer,
-                            train_path,
-                            policy_config=policy_config,
-                            exclusion_percentage=percentage,
-                            additional_arguments=additional_arguments,
-                        ),
-                        model.model_fingerprint(file_importer),
+                    train(
```
Comment (wochinge) [inline rasa/core/train.py:111]: this in turn makes `train_comparison_models` and `do_compare_training` sync, right? 🙈

File: tests/shared/importers/test_rasa.py [hunk 22]
```diff
@@ -45,13 +45,13 @@ async def test_read_conversation_tests(project: Text):
         training_data_paths=[str(Path(project) / DEFAULT_CONVERSATION_TEST_PATH)]
     )
 
-    test_stories = await importer.get_conversation_tests()
+    test_stories = importer.get_conversation_tests()
     assert len(test_stories.story_steps) == 7
 
 
 async def test_rasa_file_importer_with_invalid_config():
```
Comment (wochinge) [inline tests/shared/importers/test_rasa.py:52]: as well as `test_read_conversation_tests` and `test_rasa_file_importer_with_invalid_domain`
```suggestion
def test_rasa_file_importer_with_invalid_config():
```

File: rasa/telemetry.py [hunk 23]
```diff
@@ -698,10 +697,10 @@ async def track_model_training(
         yield  # runs the training
         return  # closes the async context
```
Comment (wochinge) [inline rasa/telemetry.py:698]:
```suggestion
        return
```

File: tests/test_validator.py [hunk 24]
```diff
@@ -8,7 +8,7 @@
 from pathlib import Path
 
 
-async def test_verify_nlu_with_e2e_story(tmp_path: Path, nlu_data_path: Path):
+def test_verify_nlu_with_e2e_story(tmp_path: Path, nlu_data_path: Path):
     print()
```
Comment (wochinge) [inline tests/test_validator.py:12]:
```suggestion
```

**Review by wochinge — COMMENTED** [review]
Looking good! 💯 Pulled in Melinda for the migration guide but in my opinion pretty fine already 👍🏻

File: docs/docs/migration-guide.mdx [hunk 25]
```diff
@@ -10,6 +10,63 @@ description: |
...
 
+## Rasa 2.x to 3.0
+
+### Training data
+
+`TrainingDataImporter` and all its implementations are updated to contain only synchronous methods.
+If you have a custom data importer or rely on some functions provided by `TrainingDataImporter`, you need
+to update your implementation and function calls.
+
+For example, this is how the data loading should look like in Rasa 3.0:
```
Comment (melindaloubser1) [inline docs/docs/migration-guide.mdx:21]:
```suggestion
For example, this is what data loading should look like in Rasa 3.0:
```

**Review by melindaloubser1 — APPROVED** [review]

File: rasa/core/train.py (hunk 21 above)
Comment (alwx) [reply rasa/core/train.py:111]: Right. Was hard to find all those little things that became sync :( 
But updated it so it should be good now!

Comment (wochinge) [reply rasa/core/train.py:111]: yeah I figured 😄 hopefully we don't decide to re-add `async` soon 😄

**Review by wochinge — APPROVED** [review]
Great work tracing all the dependent `async` usages 💯

**Pull request merged**